### ISSUE TITLE
Can't add account

### ISSUE THREAD

**Issue by vggonz** [Dimillian/IceCubesApp#16]
I can't add the account for my instance. I enter the domain "mastodon.denibol.com" but it does nothing. It's a Pleroma backend but other Mastodon clients does not have any problem loggin in. Instance metadata on https://mastodon.denibol.com/api/v1/instance

**Comment by Dimillian** [comment]
Seems like Plemora don't have configuration in the root response for this API. 

```
  public struct Configuration: Codable {
    public struct Statuses: Codable {
      public let maxCharacters: Int
      public let maxMediaAttachments: Int
    }
    
    public struct Polls: Codable {
      public let maxOptions: Int
      public let maxCharactersPerOption: Int
      public let minExpiration: Int
      public let maxExpiration: Int
    }
    
    public let statuses: Statuses
    public let polls: Polls
  }
```

I didn't look as the difference between Mastodon and Plemora yet, will probably add support for it down the line if i's close enough, but the difference on this API tell me it's not close enough.

**Comment by vggonz** [comment]
The differences between the Mastodon API and the Pleroma compatible backend are documented here: https://docs-develop.pleroma.social/backend/development/API/differences_in_mastoapi_responses/

**Comment by sbarrax** [comment]
Hi
![FE371982-C8B7-48F0-9B25-FB56873216E4](https://user-images.githubusercontent.com/1265662/212472923-4b1feb3f-8341-4610-a81a-add3612cc6cb.jpeg)
Similar problem but not the same: it says the instance is not supported

**Comment by japareaggae** [comment]
Hello, I believe the problem is that the [Instance.swift](https://github.com/Dimillian/IceCubesApp/blob/main/Packages/Models/Sources/Models/Instance.swift) model has some items that are not present in Pleroma, or other servers that implement the Mastodon API.

I found this issue since I'm also having some trouble connecting Ice Cubes to a [GoToSocial](https://github.com/superseriousbusiness/gotosocial) instance. GTS provides a Mastodon-compatible API, but with some [missing parts](https://github.com/superseriousbusiness/gotosocial/blob/main/ROADMAP.md#api-groups--endpoints). Most apps still work, though.

I did some debugging from the instance side, and I see that the app makes a request to the /api/v1/instance endpoint. Comparing the /api/v1/instance response from a GTS instance with the response from the mastodon.social instance, the missing items from the GTS response seem to be "rules" and "languages". I also tested looking at the API response from mastodon.partecipa.digital and mastodon.denibol.com, and both are also missing the "rules" array that's present in the model.

I'm not sure if there are more differences that could cause the "not supported" message to appear, since I'm not a Swift programmer and I didn't take too deep of a dive into the codebase. However, if Swift allows making items in the model optional, that could allow users to log in to those instances.

Thank you for making this app, and thanks for considering support for other server software!

**Comment by Dimillian** [comment]
![Image](https://user-images.githubusercontent.com/535509/213214822-a26928b1-95de-4994-a4a3-f221866a48fa.png)


![Image](https://user-images.githubusercontent.com/535509/213215028-42e5341b-a58f-4a94-913e-901cea725d43.png)

**Comment by japareaggae** [comment]
I probably should have mentioned my GTS instance earlier, but could you test social.quake.host as well? It's missing both the rules and languages array.

**Comment by Dimillian** [comment]
Why does Plemora want so much to do things differently? I'll take another look tomorrow and I guess I need to create a Plemora account somewhere.

**Comment by joch** [comment]
> Well, with Pleroma backend now it shows the instance info and the sign in button but it does nothing else. It seems to think for a second but nothing happen.

Just for information, it does the same on my GoToSocial instance.

**Comment by hjri** [comment]
> Why does Plemora want so much to do things differently? I'll take another look tomorrow and I guess I need to create a Plemora account somewhere.

if you want i can give you an account on post.ebin.club, i'm one of the Pleroma frontend developers. Our MastodonAPI has been implemented long long time ago and probably not on par with modern mastodon implementations.

Originally we used QvitterAPI but after some time switched to MastodonAPI + our own extensions due to... circumstances.

**Comment by japareaggae** [comment]
I downloaded Xcode and poked around the app a little, and discovered that changing vapidKey to an optional value in the [InstanceApp.swift](https://github.com/Dimillian/IceCubesApp/blob/main/Packages/Models/Sources/Models/InstanceApp.swift) model is enough to be able to log in to a GoToSocial instance. However, it leads to a Hotel California situation where I can't logout because the Settings tab calls the /api/v1/preferences endpoint, which isn't implemented yet. Still, this is better than not being able to login at all, and I'd be happy with that.

Regarding Pleroma, the server is complaining that client_name and redirect_uris are missing from the /api/v1/apps call. They are there in a `/api/v1/apps?client_name=[...]&redirect_uris=[...]` format, which I don't know if Pleroma accepts.

**Comment by Dimillian** [comment]
Maybe they accept only form data? Or maybe JSON? JSON POST would be good as I already do it for stuff like status post / edit.

**Comment by hjri** [comment]
Should accept both form data and JSON i think, but historically PleromaFE been using form data. IIRC traditionally POST requests are using parameters in the body (i.e. form data) not in the url (like GET).

You might need to also specify scopes, too. This is one of the most pain points to us because scopes tend to change from time to time as new features are added.

here's a snippet of relevant code from PleromaFE:

```js
  form.append('client_name', `PleromaFE_${window.___pleromafe_commit_hash}_${(new Date()).toISOString()}`)
  form.append('redirect_uris', REDIRECT_URI)
  form.append('scopes', 'read write follow push admin')

  return window.fetch(url, {
    method: 'POST',
    body: form
  })
```

hope it helps

**Comment by Dimillian** [comment]
Yeah right now my client is doing URL encoded form, not as body. But i'll try to turn them to JSON to see what happens.

**Comment by rjp** [comment]
Yeah, Pleroma/Akkoma require POST data in the body, not the URL (which is fair since that's what the HTML spec strongly recommends.)  Just fixed this for `madon` which made the same assumption.

**Comment by agentff6600** [comment]
i can open the sign in dialog for a pixelfed instance but as soon as I authorised access the dialog closes and I am back at the sign in without the account added.

**Comment by anantshri** [comment]
Just an update. tested the latest release on GoToSocial instance. its working now.

**Comment by perigrin** [comment]
Following from @anantshri: Trying to log into my akkoma instance just spins now without moving forward.

**Comment by bruntnorton** [comment]
> Just an update. tested the latest release on GoToSocial instance. its working now.
I am running GoToSocial on docker, the latest 0.6.0 version is not supported .  Change the version to snapshot will work.

**Comment by darrinsmart** [comment]
> Just an update. tested the latest release on GoToSocial instance. its working now.

It didn't work for me - using GTS 0.6.0 git-f9e5ec9. 

The API call to `api/v1/accounts/verify_credentials` was failing as the 'sensitive' field was missing from the response. 

However GTS have fixed this at their end (superseriousbusiness/gotosocial/issues/1332) and after building from their git repo, I can log in and see my timeline.

**Comment by Dimillian** [comment]
Nice, feel like it'll be less work for me if backend actually start to be closer to Mastodon haha

**Comment by iero** [comment]
Link to pleroma open issue about this topic here: https://git.pleroma.social/pleroma/pleroma/-/issues/3057

**Comment by catgirlinspace** [comment]
Am seeing the same issue as @perigrin on https://fediverse.catgirlin.space

**Comment by dwaynebradley** [comment]
I just migrated to https://genserver.social and it doesn't work there either. Another Pleroma site. I'm hoping that things can be worked out here as IceCubes is by far my FAVORITE fediverse app. 😎

**Comment by jaschaurbach** [comment]
I hope as well that IceCubes will Support Pleroma. 

The moment I hit "Anmelden" nothing happens

 ![image](https://github.com/Dimillian/IceCubesApp/assets/2017105/9d837278-6d40-4d54-bd07-20526e7cebb7)

**Comment by z96** [comment]
Same issue with this instance (fgc.network). Click sign in, spins for a second, nothing, then stops.

![image](https://github.com/Dimillian/IceCubesApp/assets/5444997/147216c9-8cf1-484d-83cd-5ecec06dd83d)

**Comment by fyrfli** [comment]
I did some digging on this issue myself because it works for gotosocial, mastodon, and firefish, but akkoma is still a problem for me (the "continue" button just has a spinning symbol like it can't connect).

Here's what I did:

first i tried watched the reverse proxy log file to see what is being sent through from icecubes. i intended to watch the instance log file too to see what was coming through - but it  didn't get past the proxy. the response i get is this:

```
{"level":"error","ts":1691882803.5090654,"logger":"http.log.access.log4","msg":"handled request","request":{"remote_ip":"10.0.0.1","remote_port":"63745","proto":"HTTP/3.0","method":"POST","host":"dev.bkgrdclrschm.link","uri":"/api/v1/apps?client_name=IceCubesApp&redirect_uris=icecubesapp://&scopes=read%20write%20follow%20push&website=https://github.com/Dimillian/IceCubesApp","headers":{"Accept-Language":["en-GB,en;q=0.9"],"Accept-Encoding":["gzip, deflate, br"],"Accept":["*/*"],"Priority":["u=3"],"User-Agent":["IceCubesApp/1969 CFNetwork/1410.0.3 Darwin/22.6.0"]},"tls":{"resumed":false,"version":772,"cipher_suite":4867,"proto":"h3","server_name":"dev.bkgrdclrschm.link"}},"user_id":"","duration":0.004933843,"size":286,"status":400,"resp_headers":{"Content-Length":["286"],"Date":["Sat, 12 Aug 2023 23:26:42 GMT"],"Referrer-Policy":["same-origin"],"Server":["Caddy","Cowboy"],"X-Frame-Options":["DENY"],"X-Request-Id":["F3rGciCSVnI5HV8A0vIh"],"Content-Type":["application/json; charset=utf-8"],"Cache-Control":["max-age=0, private, must-revalidate"],"X-Permitted-Cross-Domain-Policies":["none"],"X-Content-Type-Options":["nosniff"],"Access-Control-Allow-Origin":["*"],"Content-Security-Policy":["upgrade-insecure-requests;style-src 'self' 'nonce-raH1Hr8EDm4Ypmd';font-src 'self';script-src 'self' 'nonce-raH1Hr8EDm4Ypmd' ;connect-src 'self' https://dev.bkgrdclrschm.link wss://dev.bkgrdclrschm.link;media-src 'self' https:;img-src 'self' data: blob: https:;default-src 'none';base-uri 'none';frame-ancestors 'none';manifest-src 'self';"],"Access-Control-Allow-Credentials":["true"],"Permissions-Policy":["interest-cohort=()"],"X-Xss-Protection":["0"],"Access-Control-Expose-Headers":["Link,X-RateLimit-Reset,X-RateLimit-Limit,X-RateLimit-Remaining,X-Request-Id,Idempotency-Key"]}}
```

there a few other errors as well - i captured them all [here](https://paste.fyrf.li/?183eade6d30b35cd#66iryiy8vJq2FkW2EMze2LHbzvwZujCBNeBTByVdttwf) (that link expires in a week).

so then i tried to look at the /api/v1/instance data but didn't find any differences that seemed important -(those records are [here](https://paste.fyrf.li/?8c3e6c27d6709294#51v9wMmVDxUg5hZCq72Dymkr968Sth45zQruJMmdN4Q7) for gotosocial and [here](https://paste.fyrf.li/?1c386125ca7cd06e#HB6FLRW1ZbTTNjmXesZYQNCuWWdyAR2nS4ULTj43SCFU) for akkoma. 

then i realised that i would need to go through the code to see what steps are taken and i figured i'd post up my info here before i went there because that could take me quite a while to parse.

**Comment by rjp** [comment]
I think this might be because Akkoma (and presumably Pleroma) instances aren't sending the `short_description` entry in the `instance` output -- which in `Packages/Models/Sources/Models/Instance.swift` seems to be marked as required in the `Instance` struct if I'm not misreading.

But that's a bit odd because it's ignoring `description` (which both send and is the standard in both v1 and v2 calls) in favour of `short_description` (which only exists in the v1 call.)  Might be better to use `description` instead since that's required in both versions?

If I patch my Akkoma to return a `short_description` in the instance info, it does get to the "Sign In" button appearing but then you can't sign in for some other reason (presumably also down to other fields missing, etc.)

**Comment by fyrfli** [comment]
so we need to raise this with the akkoma devs then, yes?

what exactly should we say? cos i'm sure "you're not sending `short_description` in your sign in process" isn't going to be of much help.

**Comment by rjp** [comment]
I guess point them at https://docs.joinmastodon.org/entities/V1_Instance/#short_description and show that it's a required field in the V1 instance reply.

**Comment by 81nef** [comment]
encountered the same problem of instance not being supported.

![IMG_1564](https://github.com/Dimillian/IceCubesApp/assets/170366067/24593119-0eb5-4e16-acb4-9a3c608183ff)

**Comment by neelchauhan** [comment]
I run a Akkoma server at `social.monoxide.network`. If I enter `social.monoxide.network` into Ice Cubes on macOS 14.4.1, I get an infinite loading scroller.

On Tuba (Fedora/Flatpak), Semaphore (Firefox), and Tusky (various Android ROMs) I had no issues with Akkoma. I can use a native GNOME Mastodon client on my Linux HP Envy but not a native Mac one on my M2 Air, why?

**Comment by belatedly** [comment]
Just adding to this thread to note current states. These all integrate the Mastodon API into their code:

- GoToSocial v0.16.0+git-f1cbf6f: Working
- Iceshrimp v2023.12.9-dev-660d464c1: Working
- Sharkey v2024.6.0-dev: Half working - the instance shows as supported and takes me to the login/app authorization screen, but does nothing upon returning to the app after approval 
- Akkoma v3.13.2-0-050bc74: Broken. Shows a message saying “This instance is not currently supported” with the spinning sign in button

**Comment by b00mi** [comment]
After update to ios18 I can't sign in to my pixelfed.de instance 
Instance is shown. Instance is supported. Authorization icecubeApp shown, able to click Authoriz, but after this account isn't added in ice cubes and I'm back on the 'add instance page'. 
iOS 18.0.1
Icecubes 1.11.1
Pixelfed.de 3.5.3 (compatible, 0.12.3)

It worked before!

**Comment by rjp** [comment]
Akkoma has a pending (5 months!) PR to fix this - https://akkoma.dev/AkkomaGang/akkoma/pulls/821

**Comment by rjp** [comment]
I've applied the Akkoma fix to my instance but IceCubes still won't connect - probably because it's not following the spec and is sending information as URL-encoded parameters instead of in the POST body.

**Comment by DigitalStarSys** [comment]
I know this issue has been quiet for nearly a year, but I'm happy to report that Akkoma instances connect and (almost) fully function as of 2.0.9! Even with the "instance not supported" error, you can click on the button with the loading icon and it will still allow you to log in and use your account. Some things, like push notifs and extended bios, don't work, but it's a huge start and I'd say it works great!
